src/pal/src/cruntime/misctls.cpp

```diff
@@ -150,250 +164,155 @@ See MSDN for more information.
 char * __cdecl
 _ecvt( double value, int count, int * dec, int * sign )
 {
-    CONST CHAR * FORMAT_STRING = "%.348e";
-    CHAR TempBuffer[ ECVT_MAX_BUFFER_SIZE ];
-    CPalThread *pThread = NULL;
-    LPSTR lpReturnBuffer = NULL;
-    LPSTR lpStartOfReturnBuffer = NULL;
-    LPSTR lpTempBuffer = NULL;
-    LPSTR lpEndOfTempBuffer = NULL;
-    INT nTempBufferLength = 0;
-    CHAR ExponentBuffer[ 6 ];
-    INT nExponentValue = 0;
-    INT LoopIndex = 0;
-
     PERF_ENTRY(_ecvt);
     ENTRY( "_ecvt( value=%.30g, count=%d, dec=%p, sign=%p )\n",
            value, count, dec, sign );
+    
+    _ASSERTE(dec != nullptr && sign != nullptr);
+    CPalThread *pThread = InternalGetCurrentThread();
+    LPSTR lpStartOfReturnBuffer = pThread->crtInfo.ECVTBuffer;
 
-    /* Get the per-thread buffer from the thread structure. */
-    pThread = InternalGetCurrentThread();
-
-    lpStartOfReturnBuffer = lpReturnBuffer = pThread->crtInfo.ECVTBuffer;
-
-    /* Sanity checks */
-    if ( !dec || !sign )
+    if (count > ECVT_MAX_COUNT_SIZE)
     {
-        ERROR( "dec and sign have to be valid pointers.\n" );
-        *lpReturnBuffer = '\0';
-        goto done;
-    }
-    else
-    {
-        *dec = *sign = 0;
+        count = ECVT_MAX_COUNT_SIZE;
     }
 
-    if ( value < 0.0 )
-    {
-        *sign = 1;
-    }
+    const int SCALE_NAN = (int)0x80000000;
+    const int SCALE_INF = 0x7FFFFFFF;
 
-    if ( count > ECVT_MAX_COUNT_SIZE )
+    if (GetExponent(value) == 0x7ff)
     {
-        count = ECVT_MAX_COUNT_SIZE;
+        *dec = GetMantissa(value) != 0 ? SCALE_NAN : SCALE_INF;
+        *sign = GetSign(value);
+        lpStartOfReturnBuffer[0] = '\0';
+        return lpStartOfReturnBuffer;
     }
 
-    /* Get the string to work with. */
-    sprintf_s( TempBuffer, sizeof(TempBuffer), FORMAT_STRING, value );
+#ifndef __APPLE__
+    // OSX doesn't support evct_r. it support evct but we cannot use it either as it is not thread safe
+    ecvt_r(value, count, dec, sign, lpStartOfReturnBuffer, ECVT_MAX_BUFFER_SIZE);
+    return lpStartOfReturnBuffer;
+#else // __APPLE__
+
+    CHAR TempBuffer[ECVT_MAX_BUFFER_SIZE];
+   
+    *dec = *sign = 0;
 
-    /* Check to see if value was a valid number. */
-    if ( strcmp( "NaN", TempBuffer ) == 0 || strcmp( "-NaN", TempBuffer ) == 0 )
+    if (value < 0.0)
     {
-        TRACE( "value was not a number!\n" );
-        if (strcpy_s( lpStartOfReturnBuffer, ECVT_MAX_BUFFER_SIZE, "1#QNAN0" ) != SAFECRT_SUCCESS)
-        {
-            ERROR( "strcpy_s failed!\n" );
-            *lpStartOfReturnBuffer = '\0';
-            goto done;
-        }
-
-        *dec = 1;
-        goto done;
+        *sign = 1;
     }
-
-    /* Check to see if it is infinite. */
-    if ( strcmp( "Inf", TempBuffer ) == 0 || strcmp( "-Inf", TempBuffer ) == 0 )
+    
+    if (value == 0.0)
```

Comment:
It's a trivial check to fix it (just use the `*sign = GetSign(value)` call instead of the `if (value < 0.0)` check) and I'd rather see it fixed so that we can actually differentiate -0.0 from +0.0 in the future (https://github.com/dotnet/coreclr/issues/10290).